proxy/envoy/config.go

```diff
@@ -116,172 +114,116 @@ func Generate(context *ProxyContext) *Config {
-					}
-				}
-			}
-		}
-	}
+// buildHTTPListener constructs a listener for the network interface address and port
+// Use "0.0.0.0" IP address to listen on all interfaces
+// RDS parameter controls whether to use RDS for the route updates.
+func buildHTTPListener(routeConfig *HTTPRouteConfig, ip string, port int, rds bool) *Listener {
+	filters := buildFaultFilters(routeConfig)
+
+	filters = append(filters, HTTPFilter{
+		Type:   "decoder",
+		Name:   "router",
+		Config: FilterRouterConfig{},
+	})
 
-	// merge the two sets of HTTP route configs
-	httpRouteConfigs := make(HTTPRouteConfigs)
-	for port, httpRouteConfig := range httpInbound {
-		httpRouteConfigs[port] = httpRouteConfig
+	config := &HTTPFilterConfig{
+		CodecType:  "auto",
+		StatPrefix: "http",
+		AccessLog: []AccessLog{{
+			Path: DefaultAccessLog,
+		}},
+		Filters: filters,
 	}
-	for port, outgoing := range httpOutbound {
-		if incoming, ok := httpRouteConfigs[port]; ok {
-			// If the traffic is sent to a service that has instances co-located with the proxy,
-			// we choose the local service instance since we cannot distinguish between inbound and outbound packets.
-			// Note that this may not be a problem if the service port and its endpoint port are distinct.
-			httpRouteConfigs[port] = incoming.merge(outgoing)
-		} else {
-			httpRouteConfigs[port] = outgoing
+
+	if rds {
+		config.RDS = &RDS{
+			Cluster:         RDSName,
+			RouteConfigName: fmt.Sprintf("%d", port),
+			RefreshDelayMs:  1000,
 		}
+	} else {
+		config.RouteConfig = routeConfig
 	}
 
-	// merge the two sets of TCP route configs
-	tcpRouteConfigs := make(TCPRouteConfigs)
-	for port, tcpRouteConfig := range tcpInbound {
-		tcpRouteConfigs[port] = tcpRouteConfig
+	return &Listener{
+		Address: fmt.Sprintf("tcp://%s:%d", ip, port),
+		Filters: []*NetworkFilter{{
+			Type:   "read",
+			Name:   HTTPConnectionManager,
+			Config: config,
+		}},
 	}
-	for port, outgoing := range tcpOutbound {
-		if incoming, ok := tcpRouteConfigs[port]; ok {
-			// If the traffic is sent to a service that has instances co-located with the proxy,
-			// we choose the local service instance since we cannot distinguish between inbound and outbound packets.
-			// Note that this may not be a problem if the service port and its endpoint port are distinct.
-			tcpRouteConfigs[port] = incoming.merge(outgoing)
-		} else {
-			tcpRouteConfigs[port] = outgoing
-		}
+}
+
+// buildTCPListener constructs a listener for the TCP proxy
+func buildTCPListener(tcpConfig *TCPRouteConfig, ip string, port int) *Listener {
+	return &Listener{
+		Address: fmt.Sprintf("tcp://%s:%d", ip, port),
+		Filters: []*NetworkFilter{{
+			Type: "read",
+			Name: TCPProxyFilter,
+			Config: TCPProxyFilterConfig{
+				StatPrefix:  "tcp",
+				RouteConfig: tcpConfig,
+			},
+		}},
+	}
+}
+
+// buildOutboundListeners combines HTTP routes and TCP listeners
+func buildOutboundListeners(instances []*model.ServiceInstance, services []*model.Service,
```

Comment:
Suggestion about the code structure:
Can you make this function similar to buildInboundListeners, by putting the loop inside the function and have buildOutboundHTTPRoutes and buildHTTPListener work together?